solr/core/src/java/org/apache/solr/cli/SolrCLI.java

```diff
@@ -435,11 +503,34 @@ public static boolean exceptionIsAuthRelated(Exception exc) {
         && Arrays.asList(UNAUTHORIZED.code, FORBIDDEN.code).contains(((SolrException) exc).code()));
   }
 
+  /**
+   * Get a SolrClient for the given URL. The URL can be with or without /solr suffix, the method
+   * will normalize it and add '/solr'. To support certain unit tests, the hostContext can be
+   * overridden via the system property 'hostContext'.
+   *
+   * <p>Note that from Solr 10.0 Solr will use / as base URL to support /api and other contexts, so
+   * handling of hostContext differs between Solr 9.x and 10.x.
+   *
+   * @param solrUrl the base URL of Solr
+   * @return a SolrClient initialized with correct hostContext, default '/solr'
+   */
   public static SolrClient getSolrClient(String solrUrl) {
-    return new Http2SolrClient.Builder(solrUrl)
-        .withKeyStoreReloadInterval(-1, TimeUnit.SECONDS)
-        .withMaxConnectionsPerHost(32)
-        .build();
+    // SOLR-16824: Workaround for making SolrCloudExampleTest pass with custom host contexts
+    String hostContext = System.getProperty("hostContext", "/solr");
+
+    solrUrl = normalizeSolrUrl(solrUrl, true) + hostContext;
```

Comment:
This breaks some prior practice, when you could assume that you could pass in `http://localhost:8983/solr/myColl` and you'd be given a client for which a `client.commit()` would actually issue a commit for the collection in that URL. 

With my latest changes, the client would be initialized on `http://localhost:8983/solr`, and the collection part would be thrown away. I.e. the `PostTool` then needs to provide collection name as argument to `commit()`.

Since I cannot see any other Tool that breaks due to this, I think this change is fine.

On option would be to stick closer to what's in main, i.e. accept the full URL and "patch" the `hostContext` by replacing `/solr/` with `$hostContext` and let rest of URL be as is.

Reply:
Turns out that the `SolrCloudExampleTest` still fails with custom hostContext such as `/x/y`, since it will only try to strip `/solr` in normalize step, so in this case the url will be `http://localhost:8983/x/y/x/y`, i.e. double. Have a fix.